src/librbd/mirror/snapshot/GroupCreatePrimaryRequest.cc

```diff
@@ -0,0 +1,590 @@
+
+  auto req = mirror::snapshot::GroupPrepareImagesRequest<I>::create(
+    m_group_ioctx, m_group_id, cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY,
+    m_flags, &m_group_snap, &m_image_ctxs, &m_quiesce_requests, ctx);
+
+  req->send();
+}
+
+template <typename I>
+void GroupCreatePrimaryRequest<I>::handle_prepare_group_images(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to prepare group images '" << m_group_name
+                 << "': " << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  create_image_snaps();
+}
+
+template <typename I>
+void GroupCreatePrimaryRequest<I>::create_image_snaps() {
+  ldout(m_cct, 10) << "group name '" << m_group_name << "' group ID '"
+                   << m_group_id
+                   << "' group snap ID '" << m_group_snap.id << dendl;
+
+  auto ctx = librbd::util::create_context_callback<
+    GroupCreatePrimaryRequest<I>,
+    &GroupCreatePrimaryRequest<I>::handle_create_image_snaps>(this);
+  auto gather_ctx = new C_Gather(m_cct, ctx);
+
+  m_image_snap_ids.resize(m_image_ctxs.size(), CEPH_NOSNAP);
+
+  for (size_t i = 0; i < m_image_ctxs.size(); i++) {
+    image_snapshot_create2(m_image_ctxs[i], RBD_SNAP_CREATE_SKIP_QUIESCE,
+                           m_group_snap.id, &m_image_snap_ids[i],
+                           gather_ctx->new_sub());
+  }
+
+  gather_ctx->activate();
+}
+
+template <typename I>
+void GroupCreatePrimaryRequest<I>::handle_create_image_snaps(int r) {
+  ldout(m_cct, 10) << "r=" << r << dendl;
+
+  std::string group_header_oid = librbd::util::group_header_name(m_group_id);
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to create image snaps: "
+                 << cpp_strerror(r) << dendl;
+
+    if (m_ret_code == 0) {
+      m_ret_code = r;
+    }
+
+    ldout(m_cct, 10) << "undoing group create snapshot: " << r << dendl;
+    remove_interim_snapshots();
+    return;
+  } else {
+    for (size_t i = 0; i < m_image_ctxs.size(); i++) {
+      m_group_snap.snaps[i].snap_id = m_image_snap_ids[i];
+    }
+
+    m_group_snap.state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+    r = cls_client::group_snap_set(&m_group_ioctx, group_header_oid,
+                                   m_group_snap);
+    if (r < 0) {
+      lderr(m_cct) << "failed to update group snapshot metadata: "
+                 << cpp_strerror(r) << dendl;
+      if (m_ret_code == 0) {
+        m_ret_code = r;
+      }
+      // TODO: should remove_interim_snapshots() be called here for cleanup?
+    }
+
+    *m_snap_id = m_group_snap.id;
+  }
+
+  if (!m_quiesce_requests.empty()) {
+    notify_unquiesce();
+    return;
+  }
+
+  // TODO: figure out why unlink only during success?
+  if (m_ret_code == 0) {
+    unlink_peer_group();
+    return;
+  }
+
+  close_images();
+}
+
+template <typename I>
+void GroupCreatePrimaryRequest<I>::remove_interim_snapshots() {
+  ldout(m_cct, 10) << dendl;
+
+  int r = cls_client::group_snap_remove(
```

Comment:
This should use an async call.

Reply:
And this should be done only after all the image snaps are removed. An incomplete snap will be cleaned up in later calls to unlink_peer_group. If an image snap could not be deleted the first time, there is no way it will be deleted later.